### ISSUE TITLE
Preemptive generation of initial states

### ISSUE THREAD

**Issue by lemmy** [tlaplus/tlaplus#255]
TBD

**Comment by will62794** [comment]
On a related note, do you have any thoughts on the potential feasibility of parallelizing initial state generation? My understanding is that it is currently single threaded.

**Comment by lemmy** [comment]
Yes, initial states are generated sequentially by a single thread.  For the majority of specs, this is no issue unless you write a [paper comparing Alloy to TLA+](https://arxiv.org/abs/1603.03599), or you validate an inductive invariant.  In the second case, if one uses probabilistic choices to pick from a large set of initial states, one can parallelize by running multiple instances of TLCs. Another alternative would be to try [Apalache](https://github.com/informalsystems/apalache/), which, conceptually, appears to be the better fit (if it accepts your spec).

Rewriting/Changing TLC to generate initial states will lead to many side effects (e.g. how trace reconstruction works for specs with probabilistic choices such as `RandomElement`). Still, it would otherwise be straight-forward: Fork (`Fork-Join-Pool`) a new thread at every non-deterministic choice during the recursive descent in `Tool#getInitialStates` with a thread-safe `tlc2.tool.ModelChecker.DoInitFunctor`.

**Comment by will62794** [comment]
Got it, thanks. I was personally interested in it for inductive invariance checking. Not high priority, though.
